chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2193 @@
+            1. Turn radio off.
+            2. Connect cable to mic/spkr connector.
+            3. Make sure connector is firmly connected.
+            4. Turn radio on (volume may need to be set at 100%).
+            5. Ensure that the radio is tuned to channel with no activity.
+            6. Click OK to upload image to device."""))
+        return rp
+
+    def get_features(self):
+        rf = chirp_common.RadioFeatures()
+        rf.has_settings = True
+        rf.has_bank = False
+        rf.has_cross = True
+        rf.has_ctone = True
+        rf.has_rx_dtcs = True
+        rf.has_tuning_step = False
+        rf.has_ctone = True
+        rf.can_odd_split = True
+        rf.valid_name_length = 8
+        rf.valid_characters = self._valid_chars
+        rf.valid_skips = []
+        rf.valid_tmodes = ["", "Tone", "TSQL", "DTCS", "Cross"]
+        rf.valid_cross_modes = [
+            "Tone->Tone",
+            "DTCS->",
+            "->DTCS",
+            "Tone->DTCS",
+            "DTCS->Tone",
+            "->Tone",
+            "DTCS->DTCS"]
+        rf.valid_power_levels = TX_POWER
+        rf.valid_duplexes = ["", "-", "+", "split", "Off"]
+        rf.valid_modes = ["FM", "NFM"]
+        rf.valid_tuning_steps = STEPS
+
+        normal_bands = [self._vhf_range, self._uhf_range]
+
+        if self._mmap is None:
+            rf.valid_bands = [normal_bands[0], normal_bands[1]]
+        else:
+            rf.valid_bands = normal_bands
+        rf.memory_bounds = (1, 199)
+        return rf
+
+    def process_mmap(self):
+        self._memobj = bitwise.parse(MEM_FORMAT, self._mmap)
+
+    def validate_memory(self, mem):
+        msgs = chirp_common.CloneModeRadio.validate_memory(self, mem)
+
+        _msg_duplex2 = 'Memory location only supports "(None)" or "off"'
+        _msg_duplex3 = 'Memory location only supports "(None)", "+" or "off"'
+
+        if self._gmrs:
+            if mem.number < 1 or mem.number > 30:
+                if float(mem.freq) / 1000000 in GMRS_FREQS1:
+                    if mem.duplex not in ['', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex2))
+
+                if float(mem.freq) / 1000000 in GMRS_FREQS2:
+                    if mem.duplex not in ['', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex2))
+
+                if float(mem.freq) / 1000000 in GMRS_FREQS3:
+                    if mem.duplex not in ['', '+', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex3))
+
+        return msgs
+
+    def sync_in(self):
+        try:
+            self._mmap = _do_download(self)
+            self.process_mmap()
+
+        except Exception as e:
+            raise errors.RadioError("Failed to communicate with radio: %s" % e)
```

Comment:
If you would, pass through the `RadioError` exceptions. This makes it a lot easier to debug after the fact, and is why I want you to raise `RadioError` above. So make this (and the upload one) something like:

```
try:
    self_mmap = _do_download(self)
    self.process_mmap()
except RadioError:
    raise  # This causes it to raise straight to the UI instead of being translated
except Exception as e:
    # This is only for unexpected errors
    raise errors.RadioError("Failed to communicate with this radio: %s" % e)
```